contracts/exec-callee/src/entry.rs

```diff
@@ -3,26 +3,22 @@ use core::result::Result;
 
 // Import CKB syscalls and structures
 // https://nervosnetwork.github.io/ckb-std/riscv64imac-unknown-none-elf/doc/ckb_std/index.html
-use ckb_std::cstr_core::{cstr, CStr};
+use core::ffi::CStr;
 use core::slice::from_raw_parts;
 
 use crate::error::Error;
 
-pub fn main(argc: u64, argv: *const *const u8) -> Result<(), Error> {
-    let args = unsafe { from_raw_parts(argv, argc as usize) };
+pub unsafe fn main(
+    argc: core::ffi::c_int,
+    argv: *const *const core::ffi::c_char,
+) -> Result<(), Error> {
```

Comment:
I think mark `main` as unsafe is not a good idea, since user may call an unsafe function just like `from_raw_parts` and the the call is not marked as `unsafe`.

Reply:
This function is `unsafe` because it deferences raw pointers and is `pub`.

If it's not marked unsafe, you can have another safe rust function calling it with invalid pointers leading to UB. Safe rust should not cause UB, so this function has to be unsafe.

Reply:
An alternative design is to provide a safe `args` API, maybe similar to `std::env::{args, args_os}`, so users don't need to deal with raw pointers.

Reply:
See https://github.com/nervosnetwork/ckb-std/pull/26. WDYT?

Reply:
> If it's not marked unsafe, you can have another safe rust function calling it with invalid pointers leading to UB. Safe rust should not cause UB, so this function has to be unsafe.

I don't understand where the `UB` comes from, can you point it out explicitly?

Reply:
See https://doc.rust-lang.org/nomicon/safe-unsafe-meaning.html

> On functions, unsafe means that users of the function must check that function's documentation to ensure they are using it in a way that maintains the contracts the function requires.
>
> **No matter what, Safe Rust can't cause Undefined Behavior.**

If this function is not marked unsafe, you can have safe rust code like

```
main(2, core::ptr::null());
```

causing UB.

Reply:
What you want is probably `#[deny(unsafe_op_in_unsafe_fn)]`.